### DISCUSSION TITLE
Off-normal angle of incidence in 3D

### DISCUSSION THREAD

**Discussion by dalarev** [NanoComp/meep#2020] [Q&A] [answered]
Hi everybody,

Wondering if there is some guidance (e.g., an example I can guide off of) to define an off-normal source angle of incidence in 3D.

I've read the guidance for [1D](https://meep.readthedocs.io/en/latest/Python_Tutorials/Basics/#angular-reflectance-spectrum-of-a-planar-interface) and [2D](https://github.com/NanoComp/meep/blob/master/python/examples/pw-source.py), however I'm unclear on a few items:

- Are multiple current sources always needed to achieve an off-normal source? The 1D and 2D guidance (referenced above) seem to indicate so, but [this](http://www.simpetus.com/projects.html#meep_grating) example does not appear to use it.
- Does it make more sense to use MEEP in frequency-domain if I want broadband spectra as a function of angle of incidence?
- Not quite clear yet on how to define `amp_func`'s exp(j*dot(k,r)) when my source spans the _xy_ plane, propagates along _z_, polarized along _x_

Appreciate the advice!

**Comment by stevengj** [comment]
* No.  Suppose you have a source xy plane that is generating an oblique planewave propagating in ±z.   Then any planar current source with dependence proportional to `exp(i (kx x + ky y))` will produce planewaves whose xy components of their wavevector k match (kx,ky).  (The kz component is determined by the frequency).  However, using multiple current sources (e.g. with an eigenmode source in a Bloch-periodic simulation) will give you more control over the polarization of the generated wave.
* No.    With a pulsed source in the time domain, each pulse generates multiple angles corresponding to the different frequencies, so with a sequence of pulsed-source simulations you can get many angles and many wavelengths.  Or you can do a single pulsed-source simulation and look at the Fourier-transformed fields only at a single frequency to get a single angle and frequency at a time.  (See also #1991.)
* The source does not propagate in z, only the generated planewave does.  See the first point above.

**Comment by oskooi** [comment]
As an alternative to defining a planewave using the amplitude function property of the `Source` object which generates a forward- and backward-propagating planewave, you can launch an incident planewave in a *single* direction  using the `EigenModeSource` feature as demonstrated in https://meep.readthedocs.io/en/latest/Python_Tutorials/Eigenmode_Source/#planewaves-in-homogeneous-media. This can be useful for scattering calculations involving the reflectance in which the field monitor is positioned behind the source. (Nevertheless, until #1956 is resolved, you would still need to do a reference calculation with just the source to compute the input power of just the source.)

**Comment by dalarev** [comment]
Thanks very much for the insight. `EigenModeSource` looks like a useful tool I hadn't used.

In my OP I noted I wanted "broadband spectra as a function of angle of incidence", but actually I really only need broadband reflection spectra for a handful of specific angles of incidence. 

In this case, based on what I've read so far (admittedly I haven't fully grasped all the technical details), it seems I'll have to run time-stepped simulations for each angle of incidence for each frequency of interest - so, (# of angles of incidence) * (# of wavelengths) simulations. That's fine; hopefully specifying a small `df` around `fcen` means a short runtime per sim.

Briefly looking into the frequency-domain solver, I'm seeing that perhaps the `ContinuousSource` required by the frequency-domain algorithm is incompatible with `add_flux` I would need to extract the spectra?

edit: Actually, I see `add_flux` can be used with `df=0` and `nfreq=1` for this purpose.

**Comment by Nikolaos-Matthaiakakis** [comment]
So I created a file with examples based on the 2D simulation (Also including the original meep example) https://github.com/Nikolaos-MAtthaiakakis/share/blob/main/angle_example.ipynb . I also do the same thing for 3D, and also show what happens in the 2D simulation when the system is periodic along x. In the last case we can see interference from the source that spans the Y direction in both sides of the simulation domain. I imagine the use of eigenmode sources would solve this issue since it only emits from one side https://meep.readthedocs.io/en/latest/Python_Tutorials/Eigenmode_Source/ (although for broadband simulations this might increase runtime as described in the documents). The same would be true if we apply this for the 3D simulation with periodic boundaries. More experienced users can verify if my file is a correct example and the things I mention are true, or if there are better ways to avoid this issue when using periodic boundaries.

If you manage to solve your issue I would be interested to see your method since I will also require to run this kind of simulations soon

**Comment by stevengj** [comment]
@Nikolaos-MAtthaiakakis, there are two cases:

1. if the simulation is completely surrounded by PML as in your example notebook, you don't need to change the angle of the source — you can just rotate the object.  In this case you should extend the planewave current source into the PML as documented here: https://meep.readthedocs.io/en/latest/Perfectly_Matched_Layer/#planewave-sources-extending-into-pml
2. If you have a planewave incident on a periodic surface, you should use Bloch-periodic boundary conditions, and only have PML in the orthogonal direction.  (In this case you might use an eigenmode source.)

**Reply by stevengj** [reply]
Moreover, if you want to have sources that completely surround a region and create a desired incident field in that region, e.g. an incident planewave, then you should generally use an "equivalent current" source as [described in this review article](https://arxiv.org/abs/1301.5366).    

This is *not* what you did in the notebook linked above, because you used only `Ez` sources (i.e. electric currents in the z direction), whereas the principle of equivalence dictates that you should have both electric and magnetic current sources.

**Reply by Nikolaos-Matthaiakakis** [reply]
Thank you for the information. I understand that the use of the equivalent current source is required for the case of periodic boundaries (but as an alternative the eigenmode sources can be used since they only emit in one direction) . In the case where the simulation domain is terminated by PML regions the mirror flip of the emitted plane wave of electric current sources is either way absorbed by the PML so simply using current sources can be sufficient. Is this the correct understanding?

**Reply by stevengj** [reply]
No, I don't think that's correct.  It's not just a matter of the PML absorbing the outgoing waves; you also need to ensure that you get the same fields inside the box.

A mirror flip argument works for an *infinite* planar current source (e.g. one which extends all the way to the edge of the cell in a periodic simulation): in that case, the equivalent current source only emits to one side of the plane, so you can add it to its mirror image to cancel the magnetic currents while still emitting the same field on one side.  (See e.g. the discussion after equation (1) in [this paper](https://arxiv.org/pdf/1808.04215.pdf).)

In contrast, consider the equivalent current source that emits a planewave into an enclosed finite box as in your example.   If you consider only one side of the box at a time, then any given side corresponds to a *finite* sheet of electric and magnetic currents, which emits waves on *both* sides of the sheet, so you can't simply add the sheet to its mirror image to cancel the magnetic currents while still getting the same fields on the inside of the box.   Nor is there any symmetry you can apply to the box as a whole to cancel the magnetic currents.

(It *approximately* works: a finite equivalent-current sheet that is many wavelengths in diameter *mostly* emits on one side of the sheet.  That's why your enclosed-box fields look *mostly* correct, but there are slight distortions that I expect won't go away even if you refine the discretization arbitrarily.)

**Reply by Nikolaos-Matthaiakakis** [reply]
Thank you, I think I now better understand the issue.

**Comment by Nikolaos-Matthaiakakis** [comment, marked as answer]
I uploaded an example file that also includes the 2D and 3D periodic simulations for any angle of incidence.

Edit: to include broadband pulse example.

https://github.com/Nikolaos-MAtthaiakakis/share/blob/main/angle_test.ipynb 

https://user-images.githubusercontent.com/58590846/162025499-36435d3c-519d-427d-95c8-adb65d6749a9.mp4

**Reply by dalarev** [reply]
Thanks @Nikolaos-MAtthaiakakis, your codes helped me get my setup up and running.

My current problem only varies the angle of incidence in the xy-plane, so the code was a bit more straightforward. I've uploaded a MWE in 3D [here](https://github.com/dalarev/MEEP_scripts/blob/main/test_offnormal_source.ipynb) for reference.

EigenMode sources looked interesting, but I wasn't able to figure out exactly how or why to define certain parity parameters.

I'm also still not clear on why you would use two current sources (e.g., your first example in your linked code) if one seems to suffice.

Thanks again everybody for your inputs.

edit: I had been looking for a while how to plot transient fields; cheers for adding that in!
